src/sets/signerSet.ts

```diff
@@ -11,79 +12,75 @@ import { hashCompare } from '~/utils';
+      - Custody address with the higher lexicographical order wins
+      - If custody addresses have the same order (i.e. they are identical)
+        - Message with higher lexicographical hash wins
 */
-class SignerSet {
-  private _custodyAdds: Set<string>; // custody address
-  private _custodyRemoves: Set<string>; // custody address
-  private _vertexAdds: Set<string>; // pubKey
-  private _vertexRemoves: Set<string>; // pubKey
-  private _edgeAdds: Map<string, string>; // <parentKey, childKey>, hash
-  private _edgeRemoves: Map<string, string>; // <parentKey, childKey>, hash
-  private _messages: Map<string, SignerMessage>; // message hash => SignerAdd | SignerRemove
+
+export type SignerSetEvents = {
+  addCustody: (custodyAddress: string) => void;
+  removeCustody: (custodyAddress: string) => void;
+  addSigner: (signerKey: string) => void;
+  removeSigner: (signerKey: string) => void;
+};
+
+class SignerSet extends TypedEmitter<SignerSetEvents> {
+  private _custodyAdds: Map<string, CustodyAddEvent>;
+  private _custodyRemoves: Map<string, CustodyRemoveAll>;
+  private _signerAdds: Map<string, SignerAdd>;
+  private _signerRemoves: Map<string, SignerRemove>;
 
   constructor() {
-    this._custodyAdds = new Set();
-    this._custodyRemoves = new Set();
-    this._vertexAdds = new Set();
-    this._vertexRemoves = new Set();
-    this._edgeAdds = new Map();
-    this._edgeRemoves = new Map();
-    this._messages = new Map();
+    super();
+    this._custodyAdds = new Map();
+    this._custodyRemoves = new Map();
+    this._signerAdds = new Map();
+    this._signerRemoves = new Map();
   }
 
-  // TODO: add more helper functions as we integrate signer set into engine
+  getCustodyAddresses(): Set<string> {
+    return new Set([...this._custodyAdds.keys()]);
+  }
 
-  getCustodyAddresses(): string[] {
-    return Array.from(this._custodyAdds);
+  getDelegateSigners(): Set<string> {
+    return new Set([...this._signerAdds.keys()]);
   }
 
-  getDelegates(): string[] {
-    return Array.from(this._vertexAdds);
+  getAllSigners(): Set<string> {
+    return new Set([...this.getCustodyAddresses(), ...this.getDelegateSigners()]);
   }
 
-  getAllDelegates(): string[] {
-    return Array.from(this._vertices);
+  lookup(signer: string) {
+    return this.lookupCustody(signer) || this.lookupSigner(signer);
   }
 
-  sanitizeKey(key: string): string {
-    return key.toLowerCase();
+  lookupCustody(custodyAddress: string) {
+    return this._custodyAdds.get(custodyAddress);
   }
 
-  constructEdgeKey(parentKey: string, childKey: string): string {
-    return [this.sanitizeKey(parentKey), this.sanitizeKey(childKey)].toString();
+  lookupSigner(signer: string) {
+    return this._signerAdds.get(signer);
   }
 
-  deconstructEdgeKey(edgeKey: string): { parentKey: string; childKey: string } {
-    const [parentKey, childKey] = edgeKey.split(',');
-    return { parentKey, childKey };
+  sanitizeKey(key: string): string {
```

Comment:
Not a strong opinion either way, but I did it this way to future-proof it. If we want to change from lowercase to something else, we only have to change it once rather than looking at every `toLowerCase` call in the codebase to decide if it's relevant or not to sanitizing signer keys.